### DISCUSSION TITLE
[ImPlay] 分享自己写的 mpv 前端

### DISCUSSION THREAD

**Discussion by tsl0922** [hooke007/mpv_PlayKit#271] [libmpv]
曾经也是 MPV_lazy 整合包的用户，然而一直没有找到一个自己比较满意的 mpv 图形界面，于是就自己写了一个 🐶 

具体介绍请看这 2 个贴子：

- [1 个多月时间，用 C++ 撸了个跨平台桌面视频播放器](https://www.v2ex.com/t/908453)
- [尝试做了个 mpv 播放器的图形界面 - ImPlay](https://www.v2ex.com/t/900781)

不想看贴子的，可以直接到项目地址：https://github.com/tsl0922/ImPlay

欢迎提意见建议，有条件的也可以来贡献代码。

#### 截图
右键菜单和快速设置面板：
<img width="962" alt="image" src="https://user-images.githubusercontent.com/1680515/218053957-87bd7b32-d9dc-42dd-ac47-28f3af53b10b.png">
调试窗口，插件开发利器：
<img width="384" alt="image" src="https://user-images.githubusercontent.com/1680515/218059768-3df7452e-8aa3-4c58-8f9f-c2a0799742e7.png">

**Comment by hooke007** [comment]
还是希望能用wid，等待上游做d3d11/vk的支持遥遥无期。
很久以来我也没找到自己满意的前端，即便是IINA。（对我来讲render api只支持OG就是最大的毒点。
然后我发现你当前的默认context是angle...这算是og的小坑点，至少对hdr有强烈需求的用户肯定不会选择它的 https://github.com/mpv-player/mpv/issues/8530#issuecomment-772489205

**Reply by tsl0922** [reply]
没办法，windows 下 OpenGL 性能确实拉，使用 angle 翻译到 D3D 可以提升不少性能，但确实会有些功能上的损失。

设置里可以开启实验性的 wid 支持，但是还没完成。由于图形界面是自己绘制的（目前使用 OpenGL 渲染），wid 又可以配置各种渲染后端，那图形界面绘制就也得适配它们，这是不现实的。所以计划后面通过把界面绘制到自己的无边框窗口里，来实现图形界面的渲染和视频渲染各使用自己的方式。

**Reply by tsl0922** [reply]
关于 OpenGL 以外的渲染方式，等上游支持了，我会加上的。render api 的 D3D 渲染支持，社区有人做了的，但是还没合并到官方，等有空了可以试一下能不能用。

issue：https://github.com/tsl0922/ImPlay/issues/15

**Reply by hooke007** [reply]
我想知道这个前端的最终定位是什么，是 增强并贴近原生mpv体验 还是 IINA这种全面图形化（易用性与差异化）？

**Reply by tsl0922** [reply]
原生就没体验。。易用性和定制化都会做。

**Reply by tsl0922** [reply]
折腾无止境，甚至还尝试过把  imgui 的接口通过 [lua](https://github.com/tsl0922/ImPlay/tree/lua) 暴露出来，这样就可以用脚本写界面，但是 mpv 的 lua vm 没暴露给 libmpv，没法和它的 lua 插件结合起来，暂时搁置 🤣

**Reply by hooke007** [reply]
> 使用 angle 翻译到 D3D 可以提升不少性能

实际上性能倒退更多了，因为目前无法使用原生硬解，copy的性能成本对低端设备太大了
可能是bug https://github.com/mpv-player/mpv/issues/6386 & https://github.com/mpv-player/mpv/issues/7054

**Reply by tsl0922** [reply]
我这边测试 cpu 降低不少。

**Reply by tsl0922** [reply]
你提到的 issue 里说的是解码效率问题，和我上面说的（渲染）性能提升不是一个问题。

**Reply by hooke007** [reply]
默认不支持 native hwdec 对部分用户也挺致命的，这个痛点比我之前提到的hdr特性不完整支持更严重点。

**Reply by tsl0922** [reply]
mpv 默认软解。。其实现在大部分硬件软解都不是问题。这其实是 mpv 或者说 libmpv 的问题，暂时也没啥好的办法解决（wid 虽然是个办法，但是也有一些限制，不是所有平台都支持的很好）。

**Reply by hooke007** [reply]
mpv默认软件只是为了兼容并不考量性能。
就我的观察和lazy用户的反馈看，native hwdec 接近刚需了。
个人体验上讲，现实因素更复杂，我不缺性能，但是我的顶配mbp2019用软解的体验简直是灾难（烫手）

**Reply by tsl0922** [reply]
我先在默认配置里添加 `hwdec=auto`，类似的这些优化后面都可以慢慢来，只要 libmpv 支持，都可以搞。

看了下，IINA 默认是开启了 `hwdec=auto-copy`，是为了让视频滤镜默认可用吧。也测试了下 Windows，`gpu-context=angle` 时 `hwdec=auto` 确实结果是 `dxva2-copy`。

**Comment by dyphire** [comment]
除了上面提到的内核问题外（其实是最重要的），还有些其他不足：

1. 高分屏支持很差
2. 字体效果很一般
3. 不像mpv.net一样支持自定义菜单，可玩性不高

这个前端更适合入门mpv的普通用户，想深入修改和定制偏好的还是要mpv/mpv.net。尽管mpv.net停止维护，但已经是最成熟和接近原生体验的windows第三方前端了，后续自行更新mpv内核也可以一直使用下去

**Reply by tsl0922** [reply]
自定义菜单，有计划做的，mpv.net 的功能会慢慢都实现（我也是它用户）。和 mpv.net 相比，ImPlay 想做的是跨平台统一体验吧，不过确实不好搞。高分屏支持和字体效果是 imgui 这里就支持的不太好的，又不想整 qt 这种重型库来做图形界面，确实没几个选择了，后面慢慢填吧。。

**Reply by dyphire** [reply]
qt 确实太重了，不过高分屏支持和良好的字体渲染效果对于一个前端ui来说蛮重要的，希望后面能找到解决方法

**Reply by dyphire** [reply]
刚才测试字体时发现一个bug，在多台显示器上使用时在非主屏上全屏会回到主屏显示，副屏无法全屏显示

**Reply by tsl0922** [reply]
可能会有这个问题，多显示器支持还没做（代码里写死了主显示器）。

**Reply by dyphire** [reply]
测试mpv配置时又发现了其他问题
ImPlay是打开后才读取应用mpv配置，比如指定`no-border`后打开程序先显示边框后消失，这种宁可延迟响应显示也不能是这种逻辑
还有严重的问题是UI显示和`ontop`冲突，指定`ontop`将使UI无法置于最顶层，建议内部写死UI打开时自动禁用`ontop`

**Reply by tsl0922** [reply]
ontop 这个也是已知问题，还需要再研究下 imgui 的窗口创建逻辑。mpv 窗口相关的配置，都是需要再适配一遍的，这块还没怎么搞，后面按优先级慢慢填吧。

**Reply by tsl0922** [reply]
> 刚才测试字体时发现一个bug，在多台显示器上使用时在非主屏上全屏会回到主屏显示，副屏无法全屏显示

多显示器全屏问题，已经修复

**Reply by tsl0922** [reply]
> 比如指定`no-border`后打开程序先显示边框后消失，这种宁可延迟响应显示也不能是这种逻辑 

这个也已经修复，改为了延迟显示窗口。

**Reply by hooke007** [reply]
> 这个也已经修复，改为了延迟显示窗口。

Mac 下仍然是先显示窗口后消失

**Reply by tsl0922** [reply]
> Mac 下仍然是先显示窗口后消失

已修复

**Comment by tsl0922** [comment]
![image](https://user-images.githubusercontent.com/1680515/218116564-9c00dd83-544a-4aa7-90e6-4925bd88da22.png)
@dyphire 可以试下自定义字体，默认使用的是程序内置的开源字体，渲染效果确实比较差，换成微软雅黑会好很多。

**Reply by dyphire** [reply]
换字体后确实要好很多，既然如此建议内置字体改为思源字体会更好些
顺便字形范围选择似乎没什么用，我指定了思源CJK字体的中文字形结果ui实际显示的是日文字形

PS: 还是需要高分屏支持，在2K和4K屏上显示都过小了

**Reply by hooke007** [reply]
> 字形范围选择似乎没什么用

字形不选中文，界面配置中文语言就只会显示问号。

**Reply by dyphire** [reply]
> 字形不选中文，界面配置中文语言就只会显示问号。

确实，但它也实际没有使用中文字形而是用了日文字形中的汉字显示。怀疑是 imgui 自身的问题
当然也有可能和 mpv.net 一样是字符编码指定错误的原因，需要单独处理

**Reply by tsl0922** [reply]
> > 字形范围选择似乎没什么用
> 
> 字形不选中文，界面配置中文语言就只会显示问号。

是的，imgui 的字体渲染这块还比较原始，它是预先把选择范围内的文字渲染到一张图片上，需要显示某个字的时候再从图片里部分渲染（而且不支持运行时动态更改字体大小、格式等）。如果把所有的字形范围都勾上，程序会吃很多内存，看它的 milestone ，估计下个版本会支持字体的动态加载。

**Reply by tsl0922** [reply]
这些问题我也都是用久了才发现的，现在换也来不及了 😢 。。

**Reply by dyphire** [reply]
发现高分屏适配可以通过调整字体大小解决，但字体渲染这块简直无救
![image](https://user-images.githubusercontent.com/61936050/218125517-0019c404-fdad-48db-aa28-86f41e884e2d.png)
~~可以看到实际显示的是日文字形，这还好说大不了换个纯中文字形的字体显示。~~
edit: 发现使用adobe的思源CJK字体SourceHanSans-Bold可以正常显示中文字形，谷歌的不行，怪
但是注意它的英文字形还是默认的，这就很尬了，简直无救
imgui 感觉是个大坑，有种你入错坑了的感觉

**Reply by tsl0922** [reply]
哈哈，不一定，也可能是我用法不对。macOS 下我有空再测试下，虽然两个系统都有，但目前主力开发是在 win 下做的。

**Reply by dyphire** [reply]
字形范围这里添加英文的选项是不是可以接管它的英文字形显示，不知道 imgui 支不支持，现在这种字体渲染方式不能接受

**Reply by tsl0922** [reply]
> 字形范围这里添加英文的选项是不是可以接管它的英文字形显示，不知道 imgui 支不支持，现在这种字体渲染方式不能接受

英文字符显示的问题已修复，是个 bug。

**Reply by tsl0922** [reply]
> 字形不选中文，界面配置中文语言就只会显示问号。

这个问题已经解决，现在默认会把所有已知的界面显示用到的字符都加载了，但是要想显示这以外的字符还是需要到设置里开启对应语言的完整字形范围支持。暂时只能做到这样了，如果默认全加载，会拖慢程序启动速度，而且吃很多内存。

**Reply by tsl0922** [reply]
做了一个字体相关的优化：如果用户没有指定自定义字体，会根据具体显示语言使用每个操作系统下内置的推荐字体（可以在语言翻译文件里配置推荐字体位置和大小）。对于首次使用的用户来说，字体显示效果会改善很多。

**Reply by hooke007** [reply]
渲染逻辑上还是有点反直觉的地方，鼠标不动fps会降到0，导致部分操作在视觉上的反馈是明显滞后或无响应。（比如右键弹出菜单）

**Reply by tsl0922** [reply]
为了降低空闲状态下的资源占用，对图形界面 fps 是做了限制的（20fps，播放状态不受限制），也就是说输入响应反馈到界面上也会延迟 50ms（偷懒了，还没做优化，具体原因见下面描述）。

之前测试过程中发现高精度的鼠标移动会触发很多的事件，如果收到事件就立刻渲染，会导致如果不停的来回移动鼠标会触发很高频率的渲染，程序占用的资源也就会很高。

**Reply by hooke007** [reply]
> 20fps，播放状态不受限制

win是能保持10fps，所以右键的反馈只是慢而已，mac是触底0fps，右键的视觉反馈是无任何响应动画的，除非移动鼠标（才显示菜单

> 如果不停的来回移动鼠标会触发很高频率的渲染

不能强制垂直或半垂直吗

**Reply by tsl0922** [reply]
> 不能强制垂直或半垂直吗

开了垂直同步的，但是就算开了垂直同步也可以很高帧率进而吃很多资源的。这里限制的是播放器空闲状态下的帧率，播放状态下不影响的。后面优化下吧，改为实时触发事件+帧率限制（会做个设置项支持调整）。

**Reply by hooke007** [reply]
> 播放状态下不影响的。

播放时暂停也一样的（

**Reply by tsl0922** [reply]
部分操作渲染滞后的问题，复现出来了，经确认是程序逻辑上的 bug，已经修复（我这里无法明显复现了）。另外，设置里也添加了 FPS 限制选项。
